### PR TITLE
socket-collector: Improve filtering

### PR DESCRIPTION
**Pull request by blanquicet**
# socket-collector: Improve filtering

This commit improves the filtering for the socket-collector gadget introduced by PR #227. It is now possible to filter by namespace (or use -A for all-namespaces) and labels.

## Testing done

Some examples with the output in columns:

```
$ kubectl gadget socket-collector -n default
NODE      NAMESPACE    POD         PROTOCOL    LOCAL               REMOTE              STATUS
mynode    default      my-pod      TCP         10.0.0.187:44928    10.0.0.191:27017    ESTABLISHED
mynode    default      my-pod      TCP         0.0.0.0:8081        0.0.0.0:0           LISTEN
mynode    default      my-pod-2    TCP         127.0.0.1:51962     127.0.0.1:2379      ESTABLISHED
```

```
$ kubectl gadget socket-collector -n demo
NODE      NAMESPACE    POD       PROTOCOL    LOCAL            REMOTE           STATUS
mynode    demo         my-pod    UDP         10.0.0.216:68    10.0.0.107:67    ACTIVE
```

```
$ kubectl gadget socket-collector -A -p my-pod
NODE      NAMESPACE    POD       PROTOCOL    LOCAL               REMOTE              STATUS
mynode    default      my-pod    TCP         10.0.0.187:44928    10.0.0.191:27017    ESTABLISHED
mynode    default      my-pod    TCP         0.0.0.0:8081        0.0.0.0:0           LISTEN
mynode    demo         my-pod    UDP         10.0.0.216:68       10.0.0.107:67       ACTIVE
```

```
$ kubectl gadget socket-collector -n kube-system -l k8s-app=gadget
NODE      NAMESPACE      POD             PROTOCOL    LOCAL               REMOTE             STATUS
mynode    kube-system    gadget-gw7bh    TCP         10.0.0.205:57476    10.0.0.234:4240    ESTABLISHED
mynode    kube-system    gadget-gw7bh    TCP         127.0.0.1:2379      127.0.0.1:51920    ESTABLISHED
mynode    kube-system    gadget-gw7bh    TCP         127.0.0.1:2379      127.0.0.1:52010    ESTABLISHED
```

### CODE REVIEW

**Comment by blanquicet** [comment]
This PR is based on other PRs that should be reviewed first:
- [ ] https://github.com/kinvolk/inspektor-gadget/pull/238

File: pkg/gadgets/socket-collector/gadget.go [hunk 1]
```diff
@@ -72,71 +75,121 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
...
+				trace.Spec.Gadget)
 		}
-		if trace.Spec.Filter.Podname == "" {
-			gadgets.SetStatusError(trace, "TODO: Filtering only by namespace is not currently supported")
+
+		selector := gadgets.ContainerSelectorFromContainerFilter(trace.Spec.Filter)
+		filteredContainers := t.resolver.GetContainersBySelector(selector)
+		if len(filteredContainers) == 0 {
+			gadgets.SetStatusError(trace, fmt.Sprintf("Any container matched the requested filter: %+v",
+				*trace.Spec.Filter))
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/socket-collector/gadget.go:98]: "No container matches the requested filter"?

File: pkg/gadgets/socket-collector/gadget.go [hunk 2]
```diff
@@ -72,71 +75,121 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
...
-				break
+				continue
 			}
+
+			// The stored value does not matter, we are just keeping
+			// track of the visited Pods per Namespace
+			visitedPods[key] = true
+
+			log.Infof("Gadget %s: Using PID %d to retrieve network namespace of Pod %q in Namespace %q",
+				gadgetName, container.Pid, container.Podname, container.Namespace)
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/socket-collector/gadget.go:149]: It looks more like a `Debugf` than an `Infof` message to me.

File: pkg/gadgets/socket-collector/gadget.go [hunk 3]
```diff
@@ -72,71 +75,121 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
...
+			key := container.Podname + "/" + container.Namespace
+			if _, ok := visitedPods[key]; !ok {
+				if missingPods != "" {
+					missingPods += ", "
+				}
+				missingPods += key
+			}
+		}
+
+		return "", fmt.Errorf("impossible to find valid PID for %s", missingPods)
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/socket-collector/gadget.go:186]: I don't think we should pay that much attention to containers without a PID. The PID of the containers is used in other places, so if there is a container without a PID it's because there's a problem somewhere else.

File: pkg/gadgets/socket-collector/gadget.go [hunk 4]
```diff
@@ -72,71 +75,121 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
...
+		key := container.Podname + "/" + container.Namespace
+		if _, ok := visitedPods[key]; !ok {
+			totalPods++
+
+			if container.Pid == 0 {
+				discardedContainers = append(discardedContainers, container)
+
+				log.Warningf("Gadget %s: Discarding empty PID of Container %q Pod %q "+
+					"Namespace %q. Look for another Container in the same Pod",
+					gadgetName, container.Name, container.Podname, container.Namespace)
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/socket-collector/gadget.go:139]: Regarding my comment about containers with pid==0. I'd just print this warning message and that's all. I won't worry about counting containers and so on, I think it's too complicated and the benefit is very small.

**Review by mauriciovasquezbernal — COMMENTED** [review]
Some comments. I still have to try it out.

File: pkg/gadgets/socket-collector/gadget.go (hunk 4 above)
Comment (blanquicet) [reply pkg/gadgets/socket-collector/gadget.go:139]: See my comment [above](https://github.com/kinvolk/inspektor-gadget/pull/247#discussion_r700336364) 😄

File: pkg/gadgets/socket-collector/gadget.go (hunk 3 above)
Comment (mauriciovasquezbernal) [reply pkg/gadgets/socket-collector/gadget.go:186]: Wait, if there is a container with pid == 0, does the whole gadget fail?

Comment (blanquicet) [reply pkg/gadgets/socket-collector/gadget.go:186]: In the initial PR of socket-collector we decided to make the whole gadget fail (Do not show any result), if we are not able to retrieve the information for even only one of the requested Pods (i.e. there is not at least one container with pid != 0 for a given Pod). 

We made this decision because if we show the information of the rest of the Pods, it could be misunderstood as if the involved Pod does not have any socket.

Comment (mauriciovasquezbernal) [reply pkg/gadgets/socket-collector/gadget.go:186]: I think we could just make this fail if even one container has pid==0. As I said below, it's extremely rare to have a container with pid == 0 and I think we should even forbid that situation in the gadget tracer manager.  @alban do you have any opinion here?

Comment (blanquicet) [reply pkg/gadgets/socket-collector/gadget.go:186]: Understood, I will make the gadget fail if there is a container with pid == 0. BTW, this change will simplify this PR.

File: pkg/gadgets/socket-collector/gadget.go [hunk 5]
```diff
@@ -72,71 +75,82 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
...
+func start(gadgetName, node string, filteredContainers []pb.ContainerDefinition) (string, error) {
+	allSockets := []socketcollectortypes.Event{}
+
+	// Given that the socket-collector tracer works per network namespace and
+	// all the containers inside a pod/namespace share the network namespace,
+	// we only need to run the tracer with one valid PID per pod/namespace
+	visitedPods := make(map[string]bool)
+
+	for _, container := range filteredContainers {
+		key := container.Podname + "/" + container.Namespace
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/socket-collector/gadget.go:125]: Nit: usually the namespace is appended before.

File: pkg/gadgets/socket-collector/gadget.go [hunk 6]
```diff
@@ -72,71 +75,82 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
...
-			if pid != 0 {
-				break
-			}
+		selector := gadgets.ContainerSelectorFromContainerFilter(trace.Spec.Filter)
+		filteredContainers := t.resolver.GetContainersBySelector(selector)
+		if len(filteredContainers) == 0 {
+			gadgets.SetStatusError(trace, fmt.Sprintf("No container matches the requested filter: %+v",
+				*trace.Spec.Filter))
+			return
 		}
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/socket-collector/gadget.go:100]: I'm wondering how we are handling this case in the process-collector @alban.

**Review by mauriciovasquezbernal — COMMENTED** [review]
Two more comments. I'll test it once https://github.com/kinvolk/inspektor-gadget/pull/238 is merged and it's rebased.

File: pkg/gadgets/socket-collector/gadget.go (hunk 6 above)
Comment (alban) [reply pkg/gadgets/socket-collector/gadget.go:100]: The process-collector and the socket-collector works differently: they both use BPF Iterators that are specific to namespaces: to a pid namespace and to a network namespace respectively. Since network namespaces are not hierarchical, the gadget needs to enter the netns to get the data, and it needs to do that as many times as there are pods. Pid namespaces are hierarchical and the gadget pod runs with hostPID=true, so the BTF iterator is executed only one time and the filtering is done with the "containers" BPF map prepared by the gadget tracer manager. So the process-collector gadget code does not need to do anything with GetContainersBySelector etc., it is just automatic :)

Comment (mauriciovasquezbernal) [reply pkg/gadgets/socket-collector/gadget.go:100]: So if there is not any matching container it prints an empty list. I'm wondering if in that case we should also say that there is not any matching container :thinking: . For the process collector it's a actually not a big issue, if there is an empty list then there is not any matching container because at least there is one process per container. For the socket collector it's more complicated, how can the user distinguish between the case where there's not a matching container and the case there is not any open socket in the matched containers? 

In summary, I think that we should uniform the behavior of all collector when there's not a matching container even if their internal implementation is very different.

File: pkg/gadgets/socket-collector/gadget.go (hunk 5 above)
Comment (blanquicet) [reply pkg/gadgets/socket-collector/gadget.go:125]: Done. I like the code to be uniform 🤓

File: pkg/gadgets/socket-collector/gadget.go (hunk 6 above)
Comment (blanquicet) [reply pkg/gadgets/socket-collector/gadget.go:100]: I agree we should uniform them and given that, as you already said, an empty list on socket-collector could be misunderstood, I would suggest to modify the process-collector to also print an error. Do you all agree?

Comment (alban) [reply pkg/gadgets/socket-collector/gadget.go:100]: > I would suggest to modify the process-collector to also print an error. Do you all agree?

I am still hesitating about this one: if you add a check on whether the pod exists, the check is not going to be atomically done at the same time as the execution of the BPF iterator, so we could get inconsistent results.

Could that be done client-side: if kubectl-gadget receives `[]`, then it could do another call to the API server to check if the pod exists, and if not, add a message? But if the user requested `--json`, then no message?

Comment (alban) [reply pkg/gadgets/socket-collector/gadget.go:100]: I don't have a strong opinion here - adding a check client-side is also inconsistent... I'll defer to your judgment.

Comment (mauriciovasquezbernal) [reply pkg/gadgets/socket-collector/gadget.go:100]: I think it's something we can track later on.

Comment (blanquicet) [reply pkg/gadgets/socket-collector/gadget.go:100]: I created issue https://github.com/kinvolk/inspektor-gadget/issues/253 to track this.

**Comment by blanquicet** [comment]
I rebased this branch into the main now that https://github.com/kinvolk/inspektor-gadget/pull/238 was merged. PTAL and you can now start testing it. Thanks!

File: pkg/gadgets/socket-collector/gadget.go [hunk 7]
```diff
@@ -72,71 +75,82 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
...
-		trace.Spec.Filter.Namespace,
-		trace.Spec.Node,
-	)
+func start(gadgetName, node string, filteredContainers []pb.ContainerDefinition) (string, error) {
+	allSockets := []socketcollectortypes.Event{}
+
+	// Given that the socket-collector tracer works per network namespace and
+	// all the containers inside a namespace/pod share the network namespace,
+	// we only need to run the tracer with one valid PID per namespace/pod
+	visitedPods := make(map[string]bool)
```
Comment (alban) [inline pkg/gadgets/socket-collector/gadget.go:122]:
```suggestion
	visitedPods := make(map[string]struct{})
```

File: pkg/gadgets/socket-collector/gadget.go [hunk 8]
```diff
@@ -72,71 +75,82 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
...
+		if _, ok := visitedPods[key]; !ok {
+			// Make the whole gadget fail if there is a container without PID
+			// because it would be an inconsistency that has to be notified
+			if container.Pid == 0 {
+				return "", fmt.Errorf("aborting! The following container does not have PID %+v", container)
+			}
+
+			// The stored value does not matter, we are just keeping
+			// track of the visited Pods per Namespace
+			visitedPods[key] = true
```
Comment (alban) [inline pkg/gadgets/socket-collector/gadget.go:135]: This is more Go idiomatic:

```suggestion
			visitedPods[key] = struct{}{}
```

File: pkg/gadgets/socket-collector/gadget.go (hunk 7 above)
Comment (alban) [reply pkg/gadgets/socket-collector/gadget.go:122]: Note that if there are several pods with `hostNetwork=true`, then the host network namespace will be inspected several times, and the sockets will be reported several time. But I think this can be addressed in a future PR.  I am trying to address this in https://github.com/kinvolk/inspektor-gadget/pull/249

File: pkg/gadgets/socket-collector/gadget.go (hunk 8 above)
Comment (blanquicet) [reply pkg/gadgets/socket-collector/gadget.go]: TIL:
> The struct{} is a struct type with zero elements. It is often used when no information is to be stored. It has the benefit of being 0-sized, so usually no memory is required to store a value of type struct{}. [Ref](https://stackoverflow.com/a/45123114).

Applied. Thanks!

File: pkg/gadgets/socket-collector/gadget.go [hunk 9]
```diff
@@ -72,71 +75,82 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
...
-		for _, pid = range pidsInPod {
-			if pid != 0 {
-				break
-			}
+		selector := gadgets.ContainerSelectorFromContainerFilter(trace.Spec.Filter)
+		filteredContainers := t.resolver.GetContainersBySelector(selector)
+		if len(filteredContainers) == 0 {
+			gadgets.SetStatusError(trace, fmt.Sprintf("No container matches the requested filter: %+v",
+				*trace.Spec.Filter))
+			return
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/socket-collector/gadget.go:99]: Even if this test would make sense it's not the place to do it. This is the list of containers of **this** node, so if I have 10 nodes and I ask for the sockets of a given pod I'll get 9 error messages from the nodes where the pod it not scheduled.

```
$ kubectl get nodes -A
NAME           STATUS   ROLES                  AGE     VERSION
minikube       Ready    control-plane,master   10d     v1.21.2
minikube-m02   Ready    <none>                 3m44s   v1.21.2
$ kubectl get pods
NAME    READY   STATUS    RESTARTS   AGE
mypod   1/1     Running   3          10d
$ ./kubectl-gadget-linux-amd64 socket-collector -p mypod
WARN[0002] Error getting traces from node "minikube-m02": No container matches the requested filter: {Namespace:default Podname:mypod Labels:map[] ContainerName:}  args="[]" command="kubectl-gadget socket-collector"
NODE        NAMESPACE    POD      PROTOCOL    LOCAL           REMOTE       STATUS
minikube    default      mypod    TCP         0.0.0.0:5000    0.0.0.0:0    LISTEN

```

**Review by mauriciovasquezbernal — CHANGES_REQUESTED** [review]
A blocking comment regarding the warning message. I tested it and besides that comment it works fine.

File: pkg/gadgets/socket-collector/gadget.go (hunk 7 above)
Comment (mauriciovasquezbernal) [reply pkg/gadgets/socket-collector/gadget.go]: I'm wondering what should happen in that case? Ignore the pods with `hostNetwork=true`?

Comment (alban) [reply pkg/gadgets/socket-collector/gadget.go]: I think we need to have some CLI flags to say whether to include the host. It can be useful for other gadgets like execsnoop, to see what's happening on the host.

File: pkg/gadgets/socket-collector/gadget.go (hunk 9 above)
Comment (blanquicet) [reply pkg/gadgets/socket-collector/gadget.go]: I reproduced this situation in my environment. I agree that message should not be displayed. To do so, I modified the code of the socket-collector gadget to not finish the trace with an error when no container matches the requested filter, but instead, just print that message as a warning and put the trace as `Completed` with an empty output. Now, the output is clean:

```
$ ./kubectl-gadget-linux-amd64 socket-collector -A -p pod-only-on-node1
NODE     NAMESPACE      POD                  PROTOCOL    LOCAL                 REMOTE             STATUS
node1    kube-system    pod-only-on-node1    TCP         192.168.49.3:46468    10.96.0.1:443      ESTABLISHED
node1    kube-system    pod-only-on-node1    TCP         192.168.49.3:46470    10.96.0.1:443      ESTABLISHED
```

However, in the case of a non-existent pod, it will then show an empty list which, as we discussed before, could be misunderstood. Therefore, I modified the `utils/trace` in the `kubectl-gadget` side to not consider an empty output as a "successful node" (`successNodeCount`) when the `OutputMode` is different than `Stream`. Doing so, we get the following output for a non-existent pod:

```
$ ./kubectl-gadget-linux-amd64 socket-collector -A -p non-existent-pod
FATA[0000] Error getting traces from all nodes           args="[]" command="kubectl-gadget socket-collector"
```
I still would like to improve this to show the final user that the actual reason was that there were no matches with the requested filter. Right now, this info is only available in the logs of the IG running in the involved node. However, I would let that improvement for a future PR.

What do you think?

File: pkg/gadgets/socket-collector/gadget.go (hunk 7 above)
Comment (blanquicet) [reply pkg/gadgets/socket-collector/gadget.go]: Task traced by issue #258.

File: cmd/kubectl-gadget/utils/trace.go [hunk 10]
```diff
@@ -229,7 +229,12 @@ RetryLoop:
 		nodeErrors := make(map[string]string)
 		for _, i := range results.Items {
 			if i.Status.State == "Completed" || i.Status.State == "Started" {
-				successNodeCount++
+				if i.Status.Output == "" && i.Spec.OutputMode != "Stream" {
```
Comment (mauriciovasquezbernal) [inline cmd/kubectl-gadget/utils/trace.go:232]: I think this should be `i.Status.Output == "" && i.Spec.OutputMode == "Status"` to only consider the case relevant for us. If we compare against stream then we'll have to put there also other modes that we can support in the future, like `File`.

File: pkg/gadgets/socket-collector/gadget.go (hunk 9 above)
Comment (mauriciovasquezbernal) [reply pkg/gadgets/socket-collector/gadget.go]: I think it works. I put a comment regarding that implementation.

Comment (blanquicet) [reply pkg/gadgets/socket-collector/gadget.go]: I applied your suggestion.

**Review by mauriciovasquezbernal — APPROVED** [review]
LGTM!

Tested on  5.11.0-31-generic with minikube and the docker driver.

File: cmd/kubectl-gadget/utils/trace.go [hunk 11]
```diff
@@ -229,7 +229,11 @@ RetryLoop:
 		nodeErrors := make(map[string]string)
 		for _, i := range results.Items {
 			if i.Status.State == "Completed" || i.Status.State == "Started" {
-				successNodeCount++
+				if i.Status.Output == "" && i.Spec.OutputMode == "Status" {
+					contextLogger.Debugf("Ignoring empty output from node %q", i.Spec.Node)
```
Comment (alban) [inline cmd/kubectl-gadget/utils/trace.go:233]: Is this printed in case of `--json`?

File: pkg/gadgets/socket-collector/gadget.go [hunk 12]
```diff
@@ -72,71 +75,85 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
...
+			gadgets.TRACE_DEFAULT_NAMESPACE))
 		return
 	}
 
 	switch operation {
 	case "start":
-		var pid uint32 = 0
-
 		if trace.Spec.Filter == nil {
 			gadgets.SetStatusError(trace, "Missing filter")
```
Comment (alban) [inline pkg/gadgets/socket-collector/gadget.go:86]: Now that it's not mandatory to specify 1 pod in the filter, could we accept `trace.Spec.Filter == nil` as the filter that accepts everything. 

ContainerSelectorFromContainerFilter() seems to handle this correctly.

File: pkg/gadgets/socket-collector/gadget.go [hunk 13]
```diff
@@ -72,71 +75,85 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
...
-		}
-		if trace.Spec.Filter.Podname == "" {
-			gadgets.SetStatusError(trace, "TODO: Filtering only by namespace is not currently supported")
-			return
-		}
-		if len(trace.Spec.Filter.Labels) != 0 {
-			gadgets.SetStatusError(trace, "TODO: Filtering by labels is not currently supported")
-			return
-		}
 		if trace.Spec.Filter.ContainerName != "" {
```
Comment (alban) [inline pkg/gadgets/socket-collector/gadget.go:89]: If you remove the test `trace.Spec.Filter == nil` above, then 
```
if trace.Spec.Filter != nil && if trace.Spec.Filter.ContainerName != "" {
```

File: cmd/kubectl-gadget/utils/trace.go (hunk 11 above)
Comment (blanquicet) [reply cmd/kubectl-gadget/utils/trace.go:233]: It should and I think it shouldn't, even though I didn't find how to show debug logs 😅 . I think I will move that log into a comment, do you agree?

File: pkg/gadgets/socket-collector/gadget.go (hunk 12 above)
Comment (blanquicet) [reply pkg/gadgets/socket-collector/gadget.go:86]: That's true. It would be only possible if you deploy manually the trace instead of using the `kubectl-gadget`, though. 

I updated the PR, PTAL.

File: pkg/gadgets/socket-collector/gadget.go (hunk 13 above)
Comment (blanquicet) [reply pkg/gadgets/socket-collector/gadget.go:89]: Applied.

File: cmd/kubectl-gadget/utils/trace.go (hunk 11 above)
Comment (alban) [reply cmd/kubectl-gadget/utils/trace.go:233]: I think we didn't add the CLI flags for that .. 
There is an example of that at https://github.com/kinvolk/traceloop/blob/main/traceloop.go#L50
But that's outside the scope of this PR.

Your solution looks fine to me

Comment (blanquicet) [reply cmd/kubectl-gadget/utils/trace.go]: Understood. Yes, that's something we can evaluate if could be useful to be added later on. 

PR updated with the comment.

**Review by alban — APPROVED** [review]
